File: src/main/java/edu/cmu/cs/lti/ark/fn/parsing/RankedScoredRoleAssignment.java
package edu.cmu.cs.lti.ark.fn.parsing;

import com.google.common.collect.ImmutableList;
import edu.cmu.cs.lti.ark.fn.utils.DataPointWithFrameElements;
import edu.cmu.cs.lti.ark.util.ds.Range0Based;

import java.util.List;

import static edu.cmu.cs.lti.ark.fn.utils.DataPointWithFrameElements.FrameElementAndSpan;
import static edu.cmu.cs.lti.ark.util.IntRanges.xrange;
import static java.lang.Double.parseDouble;
import static java.lang.Integer.parseInt;

/**
* @author sthomson@cs.cmu.edu
*/
public class RankedScoredRoleAssignment {
	public final int rank;
	public final double score;
	public final String frame;
	public final String targetLemma;
	public final Range0Based targetSpan;
	public final String targetTokens;
	public final int sentenceIdx;
	public final List<FrameElementAndSpan> fesAndSpans;

	public RankedScoredRoleAssignment(int rank,
									  double score,
									  String frame,
									  String targetLemma,
									  Range0Based targetSpan,
									  String targetTokens,
									  int sentenceIdx,
									  List<FrameElementAndSpan> fesAndSpans) {
		this.rank = rank;
		this.score = score;
		this.frame = frame;
		this.targetLemma = targetLemma;
		this.targetSpan = targetSpan;
		this.targetTokens = targetTokens;
		this.sentenceIdx = sentenceIdx;
		this.fesAndSpans = fesAndSpans;
	}

	public static RankedScoredRoleAssignment fromPredictionLine(String frameElementsString) {
		final String[] fields = frameElementsString.split("\t");
		final int rank = parseInt(fields[0]);
		final double score = parseDouble(fields[1]);
		final int count = parseInt(fields[2]);
		final String frame = fields[3];
		final String targetLemma = fields[4];
		final String[] targetSpanStr = fields[5].split("_");
		final Range0Based targetSpan = new Range0Based(parseInt(targetSpanStr[0]), parseInt(targetSpanStr[targetSpanStr.length-1]));
		final String targetTokens = fields[6];
		final int sentenceIdx = parseInt(fields[7]);
		// Frame elements
		ImmutableList.Builder<FrameElementAndSpan> fesAndSpans = ImmutableList.builder();
		for(int i : xrange(count - 1)) {
			final String feName = fields[8+2*i];
			final Range0Based feSpan = DataPointWithFrameElements.getSpan(fields[9 + 2 * i]);
			fesAndSpans.add(new FrameElementAndSpan(feName, feSpan));
		}
		return new RankedScoredRoleAssignment(rank, score, frame, targetLemma, targetSpan, targetTokens, sentenceIdx, fesAndSpans.build());
	}
}

File: src/main/scala/edu/cmu/cs/lti/ark/fn/parsing/RankedScoredRoleAssignment.scala
package edu.cmu.cs.lti.ark.fn.parsing

import edu.cmu.cs.lti.ark.fn.utils.DataPointWithFrameElements
import edu.cmu.cs.lti.ark.util.ds.Range0Based
import RankedScoredRoleAssignment.spanToStr

/** @author sthomson@cs.cmu.edu */
case class RankedScoredRoleAssignment(rank: Int,
                                      score: Double,
                                      frame: String,
                                      targetLemma: String,
                                      targetSpan: Range0Based,
                                      targetTokens: String,
                                      sentenceIdx: Int,
                                      fesAndSpans: Array[DataPointWithFrameElements.FrameElementAndSpan]) {
  def toLine: String = {
    Seq(
      rank,
      score,
      frame,
      targetLemma,
      spanToStr("_")(targetSpan),
      targetTokens,
      sentenceIdx,
      fesAndSpans.map(fe => "%s\t%s".format(fe.name, spanToStr(":")(targetSpan))).mkString("\t")
    ).mkString("\t")
  }
}

object RankedScoredRoleAssignment {
  def fromLine(frameElementsString: String): RankedScoredRoleAssignment = {
    val fields: Array[String] = frameElementsString.split("\t")
    val fesAndSpans = for (
      Array(feName, feSpanStr) <- fields.slice(8, fields.length).grouped(2)
    ) yield {
      new DataPointWithFrameElements.FrameElementAndSpan(feName, parseSpan(":")(feSpanStr))
    }
    RankedScoredRoleAssignment(
      rank = fields(0).toInt,
      score = fields(1).toDouble,
      frame = fields(3),
      targetLemma = fields(4),
      targetSpan = parseSpan("_")(fields(5)),
      targetTokens = fields(6),
      sentenceIdx = fields(7).toInt,
      fesAndSpans = fesAndSpans.toArray
    )
  }

  def parseSpan(delimiter: String)(spanStr: String): Range0Based = {
    val spanStrs = spanStr.split(delimiter)
    new Range0Based(spanStrs(0).toInt, spanStrs(spanStrs.length - 1).toInt)
  }
  
  def spanToStr(delimiter: String)(span: Range0Based): String = "%d%s%d".format(span.start, delimiter, span.end)
}
